core/src/main/java/org/apache/iceberg/BaseTransaction.java

```diff
@@ -437,6 +430,25 @@ private void commitSimpleTransaction() {
     }
   }
 
+  private void applyUpdates(TableOperations underlyingOps) {
+    if (base != underlyingOps.refresh()) {
+      TableMetadata oldBase = base;
+      // use refreshed the metadata
+      this.base = underlyingOps.current();
+      this.current = underlyingOps.current();
+      for (PendingUpdate update : updates) {
+        // re-commit each update in the chain to apply it and update current
+        try {
+          update.commit();
+        } catch (CommitFailedException e) {
+          // fallback to old base, so that it refreshes again on retry and apply pending updates.
+          base = oldBase;
```

Comment:
@ajantha-bhat, if I understand correctly, this is happening:
1. The transaction fails to commit because of a concurrent update
2. The base is updated to the refreshed metadata to re-apply commits, but one commit fails
3. The retry runs again but detects that base is the same as the refreshed metadata, so the updates are not reapplied
4. Because the updates aren't reapplied, `commit` is called with the refreshed base and the last inner commit to succeed

Is that right?

In that case, I think that the right thing to do is to avoid catching the update's exception and throw it instead of retrying. If an update cannot be re-committed we should fail immediately rather than resetting to the original metadata to exhaust retries.

That is what we do in the REST catalog `CatalogHandlers`: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/rest/CatalogHandlers.java#L289-L294

I think that's a more correct fix.

Reply:
>  The transaction fails to commit because of a concurrent update
The base is updated to the refreshed metadata to re-apply commits, but one commit fails
The retry runs again but detects that base is the same as the refreshed metadata, so the updates are not reapplied
Because the updates aren't reapplied, commit is called with the refreshed base and the last inner commit to succeed

> Is that right?

Yes, exactly that is what happening.

> In that case, I think that the right thing to do is to avoid catching the update's exception and throw it instead of retrying. If an update cannot be re-committed we should fail immediately rather than resetting to the original metadata to exhaust retries.

Thanks for the suggestion. But I am wondering what is the point of retry configuration?
In case of single update, your suggestion holds good. But I think it will impact multiple updates.

Say I have 3 updates to apply, first update is success and the second update is failed because the first update is success. But doing a retry with rebase may help in second commit to get success right?  

There are some operations like concurrent append, which can get success on retry and some operations like concurrent schema update which cannot even success after retry. So, we may need to retry when more than one update is present and it fails right?

WDYT?

Reply:
My main worry is we are totally ignoring the retry configuration for all the scenario (as it is a common code) if I break out of the retry loop.
So, when will the retry configuration will be used then?
Let me think more on this. I am bit new to this side of the code.

Reply:
got it. `CommitFailedException` from `TableOperations.commit()` should be retired but not from the `PendingUpdate.commit()` as pending updates cannot successful even after the retry.

Reply:
Yes, exactly. So you can do the same thing as the code I referenced above.